File: src/cli/company.js
/**
 * bukio-cli — agent-first double-entry bookkeeping for Dutch SMEs.
 * Copyright (c) 2026 Erik van Kempen.
 * SPDX-License-Identifier: Apache-2.0
 */

// bukio company — show the company record, update company details.
import { t, resolveLocale } from '../i18n/index.js';
import { readFileSync, writeFileSync } from 'node:fs';
import {ensureDb, makeCtx, output, fail, table, dbError} from './util.js';
import { isValidIban } from '../core/iban.js';
import { record } from '../audit/index.js';

const LOGO_MAX_BYTES = 1_000_000; // 1 MB
const LOGO_MAX_DIMENSION = 2048;  // px, width AND height

/** Parse PNG/JPEG dimensions from the file header (SVG from width/height/viewBox). */
function logoDimensions(mime, bytes) {
  if (mime === 'image/png' && bytes.length >= 24) {
    return { width: bytes.readUInt32BE(16), height: bytes.readUInt32BE(20) };
  }
  if (mime === 'image/jpeg') {
    let i = 2;
    while (i + 9 < bytes.length) {
      if (bytes[i] !== 0xff) { i++; continue; }
      const marker = bytes[i + 1];
      if (marker >= 0xc0 && marker <= 0xcf && ![0xc4, 0xc8, 0xcc].includes(marker)) {
        return { width: bytes.readUInt16BE(i + 7), height: bytes.readUInt16BE(i + 5) };
      }
      i += 2 + bytes.readUInt16BE(i + 2);
    }
    return null;
  }
  if (mime === 'image/svg+xml') {
    const head = bytes.subarray(0, 4096).toString('utf8');
    const viewBox = head.match(/viewBox=["']\s*[\d.-]+\s+[\d.-]+\s+([\d.]+)\s+([\d.]+)\s*["']/);
    if (viewBox) return { width: Math.ceil(Number(viewBox[1])), height: Math.ceil(Number(viewBox[2])) };
    const w = head.match(/width=["']\s*([\d.]+)/);
    const h = head.match(/height=["']\s*([\d.]+)/);
    if (w && h) return { width: Math.ceil(Number(w[1])), height: Math.ceil(Number(h[1])) };
    return null; // no parsable dimensions — accept, renders at natural size
  }
  return null;
}

/**
 * Validate a logo file: PNG / JPEG / SVG only, max 1 MB, max 2048×2048 px.
 * Returns { mime, bytes, width, height } or throws LOGO_* errors.
 */
function readLogo(file) {
  let bytes;
  try {
    bytes = readFileSync(file);
  } catch (err) {
    if (err.code === 'ENOENT') {
      throw Object.assign(new Error(`logo file '${file}' not found`), { code: 'LOGO_FILE_NOT_FOUND' });
    }
    throw err;
  }
  if (bytes.length > LOGO_MAX_BYTES) {
    throw Object.assign(new Error(`logo file is ${bytes.length} bytes — the maximum is ${LOGO_MAX_BYTES}`), { code: 'LOGO_TOO_LARGE' });
  }
  let mime = null;
  if (bytes.length >= 8 && bytes[0] === 0x89 && bytes[1] === 0x50 && bytes[2] === 0x4e && bytes[3] === 0x47) {
    mime = 'image/png';
  } else if (bytes.length >= 3 && bytes[0] === 0xff && bytes[1] === 0xd8 && bytes[2] === 0xff) {
    mime = 'image/jpeg';
  } else {
    // scan a wide window: XML declarations and comment blocks push <svg deep
    const head = bytes.subarray(0, 4096).toString('utf8').replace(/^\uFEFF/, '').trimStart();
    const svgish = head.startsWith('<svg') || (head.startsWith('<?xml') && head.includes('<svg'));
    if (svgish) mime = 'image/svg+xml';
  }
  if (!mime) {
    throw Object.assign(new Error('unsupported logo format — use PNG, JPEG or SVG'), { code: 'LOGO_UNSUPPORTED_FORMAT' });
  }
  const dims = logoDimensions(mime, bytes);
  if (dims && (dims.width > LOGO_MAX_DIMENSION || dims.height > LOGO_MAX_DIMENSION)) {
    throw Object.assign(
      new Error(`logo is ${dims.width}×${dims.height} px — the maximum is ${LOGO_MAX_DIMENSION}×${LOGO_MAX_DIMENSION}`),
      { code: 'LOGO_DIMENSIONS_TOO_LARGE' },
    );
  }
  return { mime, bytes, width: dims?.width ?? null, height: dims?.height ?? null };
}

const COMPANY_FIELDS = [
  ['name', 'name', 'company name'],
  ['registrationId', 'registration_id', 'registration id (KVK for NL)'],
  ['taxId', 'tax_id', 'tax id (btw-id for NL)'],
  ['iban', 'iban', 'bank account (IBAN)'],
  ['address', 'address', 'street address (for compliant invoices)'],
  ['postalCode', 'postal_code', 'postal code'],
  ['city', 'city', 'city'],
];

function serializeCompany(row) {
  return {
    id: row.id, name: row.name, registration_id: row.registration_id, legal_form: row.legal_form,
    tax_id: row.tax_id, iban: row.iban, address: row.address,
    postal_code: row.postal_code, city: row.city, vat_module: row.vat_module,
    kor_flag: row.kor_flag, fiscal_year_end: row.fiscal_year_end,
    country: row.country, base_currency: row.base_currency, locale: row.locale,
    profile_version: row.profile_version,
    logo_mime: row.logo_mime ?? null,
    logo_bytes: row.logo ? row.logo.length : null,
  };
}

function getCompany(db) {
  return db.prepare('SELECT * FROM company WHERE id = 1').get() ?? null;
}

export function make(program) {
  const company = program.command('company').description('company record: show, update');

  company
    .command('show')
    .description('show the company record')
    .action((opts, command) => {
      const ctx = makeCtx(command);
      try {
        const db = ensureDb(ctx);
        const locale = resolveLocale(ctx, db);
        try {
          const row = getCompany(db);
          if (!row) throw dbError('NO_COMPANY', 'no company — run bukio init first');
          output(ctx, { company: serializeCompany(row) }, (d) => {
            table([d.company], [
              { key: 'name', label: t('company.name', {}, locale) },
              { key: 'country', label: t('company.country', {}, locale) },
              { key: 'legal_form', label: t('company.legalForm', {}, locale) },
              { key: 'registration_id', label: t('company.regId', {}, locale) },
              { key: 'tax_id', label: t('company.taxId', {}, locale) },
              { key: 'iban', label: t('company.iban', {}, locale) },
              { key: 'address', label: t('company.address', {}, locale) },
              { key: 'postal_code', label: t('company.postalCode', {}, locale) },
              { key: 'city', label: t('company.city', {}, locale) },
              { key: 'base_currency', label: t('company.currency', {}, locale) },
              { key: 'locale', label: t('company.language', {}, locale) },
            ]);
          });
        } finally {
          db.close();
        }
      } catch (err) {
        fail(ctx, err);
      }
    });

  company
    .command('update')
    .description('update company details (address, IBAN, btw-id, name, kvk, logo)')
    .option('--name <name>', 'company name')
    .option('--registration-id <id>', 'company registration number (KVK for NL)')
    .option('--tax-id <id>', 'tax identification number (btw-id for NL)')
    .option('--kvk <kvk>', '[deprecated] alias for --registration-id')
    .option('--btw-id <id>', '[deprecated] alias for --tax-id')
    .option('--country <CC>', 'country — immutable after init (rejected)')
    .option('--iban <iban>', 'bank account (IBAN)')
    .option('--address <address>', 'street address (for compliant invoices)')
    .option('--postal-code <code>', 'postal code')
    .option('--city <city>', 'city')
    .option('--logo <file>', 'set the invoice logo (PNG/JPEG/SVG, max 1 MB, stored in the database)')
    .option('--remove-logo', 'remove the logo')
    .option('--dry-run', 'show the plan without writing')
    .action((opts, command) => {
      const ctx = makeCtx(command);
      try {
        const db = ensureDb(ctx);
        try {
          const row = getCompany(db);
          if (!row) throw dbError('NO_COMPANY', 'no company — run bukio init first');

          // deprecated aliases: --kvk -> --registration-id, --btw-id -> --tax-id
          if (opts.kvk !== undefined && opts.registrationId !== undefined) {
            opts._warnings = [...(opts._warnings ?? []), '--kvk ignored because --registration-id was also given'];
          } else if (opts.kvk !== undefined) {
            opts.registrationId = opts.kvk;
            opts._warnings = [...(opts._warnings ?? []), '--kvk is deprecated — use --registration-id'];
          }
          if (opts.btwId !== undefined && opts.taxId !== undefined) {
            opts._warnings = [...(opts._warnings ?? []), '--btw-id ignored because --tax-id was also given'];
          } else if (opts.btwId !== undefined) {
            opts.taxId = opts.btwId;
            opts._warnings = [...(opts._warnings ?? []), '--btw-id is deprecated — use --tax-id'];
          }
          // country is immutable after init (decision §9.1.5)
          if (opts.country !== undefined) {
            const want = String(opts.country).trim().toUpperCase();
            if (want !== (row.country ?? 'NL')) {
              throw Object.assign(new Error(`country is immutable after init — company stays ${row.country ?? 'NL'} (re-init a new DB for another country)`), { code: 'COUNTRY_IMMUTABLE' });
            }
          }
          const changes = {};
          for (const [opt, col, label] of COMPANY_FIELDS) {
            if (opts[opt] !== undefined) changes[col] = String(opts[opt]).trim();
          }
          let logo = null;
          if (opts.logo !== undefined) logo = readLogo(opts.logo);
          if (opts.removeLogo) logo = { mime: null, bytes: null };
          if (Object.keys(changes).length === 0 && logo === null) {
            throw Object.assign(new Error('nothing to update — pass at least one of --name/--registration-id/--tax-id/--iban/--address/--postal-code/--city/--logo/--remove-logo'), { code: 'NOTHING_TO_UPDATE' });
          }
          for (const [opt, col, label] of COMPANY_FIELDS) {
            if (changes[col] === '' && col !== 'tax_id') {
              throw Object.assign(new Error(`${label} cannot be empty`), { code: 'INVALID_VALUE' });
            }
          }
          if (changes.iban && !isValidIban(changes.iban)) {
            throw Object.assign(new Error(`invalid IBAN '${changes.iban}'`), { code: 'INVALID_IBAN' });
          }

          const plan = {
            company: { ...serializeCompany(row), ...changes },
            changes,
            logo: logo === null ? null
              : { mime: logo.mime, bytes: logo.bytes ? logo.bytes.length : 0, width: logo.width, height: logo.height },
            ...(opts._warnings?.length ? { warnings: opts._warnings } : {}),
            dryRun: true,
          };
          if (ctx.dryRun) {
            output(ctx, plan, (d) => {
              console.log('plan: company update');
              for (const [k, v] of Object.entries(d.changes)) console.log(`  ${k}: '${row[k]}' -> '${v}'`);
              if (d.logo) console.log(`  logo: ${d.logo.mime} (${d.logo.bytes} bytes${d.logo.width ? `, ${d.logo.width}×${d.logo.height} px` : ''})${d.logo.bytes === 0 ? ' — removed' : ''}`);
              for (const w of d.warnings ?? []) console.error(`warning: ${w}`);
              console.log('(dry run — nothing written)');
            });
            return;
          }

          const sets = Object.keys(changes).map((c) => `${c} = ?`).join(', ');
          if (sets) {
            db.prepare(`UPDATE company SET ${sets} WHERE id = 1`).run(...Object.values(changes));
          }
          if (logo !== null) {
            db.prepare('UPDATE company SET logo = ?, logo_mime = ? WHERE id = 1')
              .run(logo.bytes, logo.mime);
          }
          const updated = getCompany(db);
          record(db, {
            actor: ctx.actor, action: 'company.update', command: 'company update',
            args: {
              ...changes,
              ...(logo !== null ? { logo: logo.mime ? `${logo.mime} (${logo.bytes.length} bytes)` : 'removed' } : {}),
            },
            outcome: 'ok',
          });
          output(ctx, {
            company: serializeCompany(updated), changes,
            ...(opts._warnings?.length ? { warnings: opts._warnings } : {}),
          }, (d) => {
            console.log('company updated:');
            for (const [k, v] of Object.entries(d.changes)) console.log(`  ${k}: '${row[k]}' -> '${v}'`);
            if (logo !== null && logo.mime) console.log(`  logo: ${logo.mime} (${logo.bytes.length} bytes)`);
            if (logo !== null && !logo.mime) console.log('  logo: removed');
            for (const w of d.warnings ?? []) console.error(`warning: ${w}`);
          });
        } finally {
          db.close();
        }
      } catch (err) {
        fail(ctx, err);
      }
    });

  company
    .command('logo')
    .description('extract the stored logo to a file')
    .requiredOption('--out <file>', 'output file path')
    .action((opts, command) => {
      const ctx = makeCtx(command);
      try {
        const db = ensureDb(ctx);
        try {
          const row = getCompany(db);
          if (!row?.logo) throw Object.assign(new Error('no logo stored — set one with company update --logo'), { code: 'LOGO_NOT_SET' });
          writeFileSync(opts.out, row.logo);
          output(ctx, { out: opts.out, mime: row.logo_mime, bytes: row.logo.length }, (d) => {
            console.log(`logo written: ${d.out} (${d.mime}, ${d.bytes} bytes)`);
          });
        } finally {
          db.close();
        }
      } catch (err) {
        fail(ctx, err);
      }
    });
}

File: src/company.rs
// bukio-cli — agent-first double-entry bookkeeping for SMEs.
// Copyright (c) 2026 Erik van Kempen.
// SPDX-License-Identifier: Apache-2.0
//
// Company settings (mirrors src/cli/company.js): show, update, logo.

use crate::audit::{record, RecordArgs};
use crate::dates::is_valid_iban;
use crate::money::{BukioError, Result};
use rusqlite::Connection;
use serde_json::{json, Value};

const COMPANY_FIELDS: &[(&str, &str, &str)] = &[
    ("name", "name", "company name"),
    ("registration-id", "registration_id", "registration id"),
    ("tax-id", "tax_id", "tax id"),
    ("iban", "iban", "bank account (IBAN)"),
    ("address", "address", "street address"),
    ("postal-code", "postal_code", "postal code"),
    ("city", "city", "city"),
];

/// Serialize a company row to JSON (without the logo blob).
/// Column order after migration 022: id(0) name(1) registration_id(2) legal_form(3)
/// tax_id(4) iban(5) vat_module(6) kor_flag(7) fiscal_year_end(8) created_at(9)
/// updated_at(10) address(11) postal_code(12) city(13) logo(14) logo_mime(15)
/// country(16) base_currency(17) locale(18) profile_version(19)
pub fn serialize_company(row: &rusqlite::Row<'_>) -> rusqlite::Result<Value> {
    let logo_bytes: Option<Vec<u8>> = row.get(14)?;
    Ok(json!({
        "id": row.get::<_, i64>(0)?,
        "name": row.get::<_, String>(1)?,
        "registration_id": row.get::<_, Option<String>>(2)?,
        "legal_form": row.get::<_, Option<String>>(3)?,
        "tax_id": row.get::<_, Option<String>>(4)?,
        "iban": row.get::<_, Option<String>>(5)?,
        "address": row.get::<_, Option<String>>(11)?,
        "postal_code": row.get::<_, Option<String>>(12)?,
        "city": row.get::<_, Option<String>>(13)?,
        "vat_module": row.get::<_, i64>(6)?,
        "kor_flag": row.get::<_, i64>(7)?,
        "fiscal_year_end": row.get::<_, Option<String>>(8)?,
        "country": row.get::<_, Option<String>>(16)?,
        "base_currency": row.get::<_, Option<String>>(17)?,
        "locale": row.get::<_, Option<String>>(18)?,
        "profile_version": row.get::<_, i64>(19)?,
        "logo_mime": row.get::<_, Option<String>>(15)?,
        "logo_bytes": logo_bytes.as_ref().map(|b| b.len() as i64),
    }))
}

pub fn get_company(db: &Connection) -> Result<Value> {
    let row = db
        .query_row("SELECT * FROM company WHERE id = 1", [], |r| {
            serialize_company(r)
        })
        .map_err(|_| BukioError::new("NO_COMPANY", "no company — run bukio init first"))?;
    Ok(row)
}

/// Update company fields. Returns the updated company and a map of changes.
pub fn update_company(
    db: &Connection,
    changes: &[(String, String)], // (column, value) pairs
    logo_bytes: Option<Vec<u8>>,
    logo_mime: Option<&str>,
    actor: &str,
) -> Result<(Value, Value)> {
    let row = get_company(db)?;

    // validate each change
    let mut updates: Vec<(String, String)> = Vec::new();
    for (col, val) in changes {
        if val.is_empty() && col != "tax_id" {
            let label = COMPANY_FIELDS
                .iter()
                .find(|(opt, c, _)| *c == *col)
                .map(|f| f.2)
                .unwrap_or("field");
            return Err(BukioError::new(
                "INVALID_VALUE",
                format!("{label} cannot be empty"),
            ));
        }
        if col == "iban" && !val.is_empty() && !is_valid_iban(val) {
            return Err(BukioError::new(
                "INVALID_IBAN",
                format!("invalid IBAN '{val}'"),
            ));
        }
        updates.push((col.clone(), val.clone()));
    }

    if updates.is_empty() && logo_bytes.is_none() && logo_mime.is_none() {
        return Err(BukioError::new("NOTHING_TO_UPDATE", "nothing to update"));
    }

    // apply text field changes
    if !updates.is_empty() {
        let set_clauses: Vec<String> = updates.iter().map(|(c, _)| format!("{c} = ?")).collect();
        let sql = format!("UPDATE company SET {} WHERE id = 1", set_clauses.join(", "));
        let params: Vec<String> = updates.iter().map(|(_, v)| v.clone()).collect();
        let param_refs: Vec<&dyn rusqlite::types::ToSql> = params
            .iter()
            .map(|p| p as &dyn rusqlite::types::ToSql)
            .collect();
        db.execute(&sql, param_refs.as_slice()).map_err(sql_err)?;
    }

    // apply logo change: Some(bytes) sets it, Some(empty) clears it
    if logo_bytes.is_some() || logo_mime.is_some() {
        let clear = logo_bytes.as_ref().map(|b| b.is_empty()).unwrap_or(false);
        let (store_bytes, store_mime): (Option<Vec<u8>>, Option<String>) = if clear {
            (None, None)
        } else {
            (logo_bytes.clone(), logo_mime.map(|m| m.to_string()))
        };
        db.execute(
            "UPDATE company SET logo = ?1, logo_mime = ?2 WHERE id = 1",
            rusqlite::params![store_bytes, store_mime],
        )
        .map_err(sql_err)?;
    }

    let updated = get_company(db)?;

    // build changes map for audit
    let changes_map = serde_json::Map::from_iter(
        updates
            .iter()
            .map(|(k, v)| (k.clone(), Value::String(v.clone()))),
    );
    let changes_map = Value::Object(changes_map);

    record(
        db,
        RecordArgs {
            actor,
            action: "company.update",
            command: Some("company update"),
            args: Some(changes_map.clone()),
            outcome: "ok",
            entry_ids: vec![],
        },
    )?;

    Ok((updated, changes_map))
}

/// Read + validate a logo file: PNG/JPEG/SVG only, max 1 MB, max 2048×2048 px
/// (mirrors JS readLogo in cli/company.js). Returns (bytes, mime).
pub fn read_logo_file(file: &str) -> Result<(Vec<u8>, String)> {
    const MAX_BYTES: usize = 1_000_000;
    const MAX_DIM: i64 = 2048;
    let bytes = std::fs::read(file).map_err(|_| {
        BukioError::new(
            "LOGO_FILE_NOT_FOUND",
            format!("logo file '{file}' not found"),
        )
    })?;
    if bytes.len() > MAX_BYTES {
        return Err(BukioError::new(
            "LOGO_TOO_LARGE",
            format!(
                "logo file is {} bytes — the maximum is {MAX_BYTES}",
                bytes.len()
            ),
        ));
    }
    let mime = if bytes.len() >= 8
        && bytes[0] == 0x89
        && bytes[1] == 0x50
        && bytes[2] == 0x4e
        && bytes[3] == 0x47
    {
        "image/png"
    } else if bytes.len() >= 3 && bytes[0] == 0xff && bytes[1] == 0xd8 && bytes[2] == 0xff {
        "image/jpeg"
    } else {
        // wide window: XML declarations and comment blocks push <svg deep
        let head = String::from_utf8_lossy(&bytes[..bytes.len().min(4096)])
            .trim_start_matches('\u{feff}')
            .trim_start()
            .to_string();
        if head.starts_with("<svg") || (head.starts_with("<?xml") && head.contains("<svg")) {
            "image/svg+xml"
        } else {
            return Err(BukioError::new(
                "LOGO_UNSUPPORTED_FORMAT",
                "unsupported logo format — use PNG, JPEG or SVG",
            ));
        }
    };
    if let Some((w, h)) = logo_dimensions(mime, &bytes) {
        if w > MAX_DIM || h > MAX_DIM {
            return Err(BukioError::new(
                "LOGO_DIMENSIONS_TOO_LARGE",
                format!("logo is {w}×{h} px — the maximum is {MAX_DIM}×{MAX_DIM}"),
            ));
        }
    }
    Ok((bytes, mime.to_string()))
}

fn logo_dimensions(mime: &str, bytes: &[u8]) -> Option<(i64, i64)> {
    if mime == "image/png" && bytes.len() >= 24 {
        return Some((
            i64::from(u32::from_be_bytes([
                bytes[16], bytes[17], bytes[18], bytes[19],
            ])),
            i64::from(u32::from_be_bytes([
                bytes[20], bytes[21], bytes[22], bytes[23],
            ])),
        ));
    }
    if mime == "image/jpeg" {
        let mut i = 2usize;
        while i + 9 < bytes.len() {
            if bytes[i] != 0xff {
                i += 1;
                continue;
            }
            let marker = bytes[i + 1];
            if (0xc0..=0xcf).contains(&marker) && ![0xc4, 0xc8, 0xcc].contains(&marker) {
                let h = u16::from_be_bytes([bytes[i + 5], bytes[i + 6]]);
                let w = u16::from_be_bytes([bytes[i + 7], bytes[i + 8]]);
                return Some((i64::from(w), i64::from(h)));
            }
            let len = u16::from_be_bytes([bytes[i + 2], bytes[i + 3]]) as usize;
            i += 2 + len;
        }
        return None;
    }
    if mime == "image/svg+xml" {
        let head = String::from_utf8_lossy(&bytes[..bytes.len().min(4096)]).to_string();
        // viewBox="x y w h"
        let re =
            regex::Regex::new(r#"viewBox=["']\s*[\d.-]+\s+[\d.-]+\s+([\d.]+)\s+([\d.]+)\s*["']"#)
                .unwrap();
        if let Some(c) = re.captures(&head) {
            let w = c[1].parse::<f64>().ok()?.ceil() as i64;
            let h = c[2].parse::<f64>().ok()?.ceil() as i64;
            return Some((w, h));
        }
        let wre = regex::Regex::new(r#"width=["']\s*([\d.]+)"#).unwrap();
        let hre = regex::Regex::new(r#"height=["']\s*([\d.]+)"#).unwrap();
        if let (Some(w), Some(h)) = (wre.captures(&head), hre.captures(&head)) {
            return Some((
                w[1].parse::<f64>().ok()?.ceil() as i64,
                h[1].parse::<f64>().ok()?.ceil() as i64,
            ));
        }
        return None; // no parsable dims — accept, renders at natural size
    }
    None
}

/// Extract the stored logo.
pub fn get_logo(db: &Connection) -> Result<(Vec<u8>, String)> {
    let row = db
        .query_row(
            "SELECT logo, logo_mime FROM company WHERE id = 1 AND logo IS NOT NULL",
            [],
            |r| Ok((r.get::<_, Vec<u8>>(0)?, r.get::<_, String>(1)?)),
        )
        .map_err(|_| BukioError::new("LOGO_NOT_SET", "no logo stored"))?;
    Ok(row)
}

fn sql_err(e: rusqlite::Error) -> BukioError {
    BukioError::new("DB_ERROR", e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::open_db;

    fn db() -> Connection {
        let d = open_db(":memory:").unwrap();
        d.execute("INSERT INTO company (name) VALUES ('Test BV')", [])
            .unwrap();
        d
    }

    #[test]
    fn get_company_basic() {
        let d = db();
        let c = get_company(&d).unwrap();
        assert_eq!(c["name"], "Test BV");
    }

    #[test]
    fn update_name() {
        let d = db();
        let (updated, changes) = update_company(
            &d,
            &[("name".into(), "New BV".into())],
            None,
            None,
            "human:erik",
        )
        .unwrap();
        assert_eq!(updated["name"], "New BV");
        assert_eq!(changes["name"], "New BV");
    }

    #[test]
    fn update_empty_name_rejected() {
        let d = db();
        let err = update_company(&d, &[("name".into(), "".into())], None, None, "human:erik")
            .unwrap_err();
        assert_eq!(err.code, "INVALID_VALUE");
    }

    #[test]
    fn nothing_to_update_rejected() {
        let d = db();
        let err = update_company(&d, &[], None, None, "human:erik").unwrap_err();
        assert_eq!(err.code, "NOTHING_TO_UPDATE");
    }
}
